### DISCUSSION TITLE
LSP Zero and efm-langserver

### DISCUSSION THREAD

**Discussion by olimorris** [VonHeikemen/lsp-zero.nvim#293] [Show and tell]
Thought I'd share my LSP Zero and efm-langserver config, [here](https://github.com/olimorris/dotfiles/blob/2e5aa89ffd227d15a0d79256d3647a46a99c8f12/.config/nvim/lua/Oli/plugins/lsp.lua#L333-L356).

Worth noting that I had to manually set some server capabilities for efm in the `lsp.on_attach` method:

```lua
if client.name == "efm" then
  client.server_capabilities.documentFormattingProvider = true
  client.server_capabilities.documentFormattingRangeProvider = true
end
```

**Comment by VonHeikemen** [comment]
You think it would be worth adding some helper functions for efm? To create a config based on tools, instead of declaring tools for each language?

I can make something like this:

```lua
local lsp_zero = require('lsp-zero')

local efm_opts = lsp_zero.efm.tools({
  {
    name = 'eslint-fix',
    formatCommand = "eslint --fix '${INPUT}'",
    formatStdin = false,
    languages = {'javascript', 'typescript'},
  },
  {
    name = 'prettierd',
    formatCommand = "prettierd '${INPUT}' ${--range-start=charStart} ${--range-end=charEnd}",
    formatCanRange = true,
    formatStdin = true,
    rootMarkers = {'.prettierrc.json', '.prettierrc.js'},
    languages = {'javascript', 'typescript'},
  },
  {
    name = 'stylua',
    formatCommand = 'stylua --color Never ${--range-start:charStart} ${--range-end:charEnd} -',
    formatCanRange = false,
    formatStdin = true,
    rootMarkers = {'stylua.toml', '.stylua.toml'},
    languages = {'lua'}
  },
})

require('lspconfig').efm.setup(efm_opts)

lsp_zero.format_mapping('gq', {
  servers = {
    efm = lsp_zero.efm.langs({with = 'formatCommand'})
  },
})
```

I know this will only be useful for those who hate yaml. (I hate yaml configs)

**Reply by olimorris** [reply]
That would be a really nice quality of life improvement in using efm imo!

**Reply by VonHeikemen** [reply]
@olimorris awesome. I've created a new issue if you want discuss this #344 

If you want to test this api, you can use the branch `efm-client`.
